### ISSUE TITLE
How should the gtf file for `TagReadWithGeneFunction` be structured? 

### ISSUE THREAD

**Issue by jolespin** [broadinstitute/Drop-seq#78]
### Instructions
My `gtf` is generated from `maker` and it is missing `gene id`, `gene name`, and `transcript name`.  I plan on fixing these errors manually. 

### Affected tool(s)
```
TagReadWithGeneFunction -INPUT dropseq_output/Frozen_HJC7MBGX5/merged.bam -ANNOTATIONS_FILE /usr/local/projdata/0568/projects/PLANKTON/illumina_aallen/jespinoz/References/Cylindrotheca/annotations/cylindrotheca_rd3.all.maker.gtf -OUTPUT dropseq_output/Frozen_HJC7MBGX5/star_gene_exon_tagged.bam
```
### Affected version(s)
`v2.1`

### Description 

```
-bash-4.1$ cat dropseq_output/Frozen_HJC7MBGX5/logging/star_gene_exon_tagged.bam.log
INFO    2019-01-24 02:24:17 TagReadWithGeneFunction

********** NOTE: Picard's command line syntax is changing.
**********
********** For more information, please see:
********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)
**********
********** The command line looks like this in the new syntax:
**********
**********    TagReadWithGeneFunction -INPUT dropseq_output/Frozen_HJC7MBGX5/merged.bam -ANNOTATIONS_FILE /usr/local/projdata/0568/projects/PLANKTON/illumina_aallen/jespinoz/References/Cylindrotheca/annotations/cylindrotheca_rd3.all.maker.gtf -OUTPUT dropseq_output/Frozen_HJC7MBGX5/star_gene_exon_tagged.bam
**********


02:24:19.432 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/jespinoz/Algorithms/Mapping/dropseq_pipeline/Drop-seq_tools-2.1.0/jar/lib/picard-2.18.14.jar!/com/intel/gkl/native/libgkl_compression.so
[Thu Jan 24 02:24:19 EST 2019] TagReadWithGeneFunction INPUT=dropseq_output/Frozen_HJC7MBGX5/merged.bam OUTPUT=dropseq_output/Frozen_HJC7MBGX5/star_gene_exon_tagged.bam ANNOTATIONS_FILE=/usr/local/projdata/0568/projects/PLANKTON/illumina_aallen/jespinoz/References/Cylindrotheca/annotations/cylindrotheca_rd3.all.maker.gtf    GENE_NAME_TAG=gn GENE_STRAND_TAG=gs GENE_FUNCTION_TAG=gf READ_FUNCTION_TAG=XF USE_STRAND_INFO=true VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false
[Thu Jan 24 02:24:19 EST 2019] Executing as jespinoz@dell-15.jcvi.org on Linux 2.6.32-696.18.7.el6.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_121-b15; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.1.0(e9a342e_1547567906)
[Thu Jan 24 02:24:21 EST 2019] org.broadinstitute.dropseqrna.metrics.TagReadWithGeneFunction done. Elapsed time: 0.03 minutes.
Runtime.totalMemory()=2058354688
Exception in thread "main" picard.annotation.AnnotationException: Invalid GTF line:
000000F|arrow   maker   exon    266 1689    .   +   .   transcript_id "CylT000001-RA|arrow-processed-gene-0.141-mRNA-1";
Problems:
Missing gene_id
Missing gene_name
Missing transcript_name
    at org.broadinstitute.dropseqrna.annotation.GTFParser.next(GTFParser.java:97)
    at org.broadinstitute.dropseqrna.annotation.GTFParser.next(GTFParser.java:39)
    at htsjdk.samtools.util.PeekableIterator.advance(PeekableIterator.java:71)
    at htsjdk.samtools.util.PeekableIterator.<init>(PeekableIterator.java:38)
    at org.broadinstitute.dropseqrna.utils.FilteredIterator.<init>(FilteredIterator.java:37)
    at org.broadinstitute.dropseqrna.annotation.GTFReader$FilteringGTFParser.<init>(GTFReader.java:109)
    at org.broadinstitute.dropseqrna.annotation.GTFReader$FilteringGTFParser.<init>(GTFReader.java:107)
    at org.broadinstitute.dropseqrna.annotation.GTFReader.load(GTFReader.java:74)
    at org.broadinstitute.dropseqrna.annotation.GTFReader.load(GTFReader.java:70)
    at org.broadinstitute.dropseqrna.annotation.GeneAnnotationReader.loadGTFFile(GeneAnnotationReader.java:67)
    at org.broadinstitute.dropseqrna.annotation.GeneAnnotationReader.loadAnnotationsFile(GeneAnnotationReader.java:51)
    at org.broadinstitute.dropseqrna.metrics.TagReadWithGeneFunction.doWork(TagReadWithGeneFunction.java:113)
    at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:295)
    at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:103)
    at org.broadinstitute.dropseqrna.cmdline.DropSeqMain.main(DropSeqMain.java:42)
```
#### Steps to reproduce
Creating a `gff` file from `maker` and using cufflink's `gffread` to convert to `gtf`

#### Expected behavior
I expected it to work with transcript id but the gff to gtf conversion did not work well.

#### Actual behavior
It didn't because I was missing the aformentioned fields. 

Here is the top few lines of my gff file:
```
-bash-4.1$ head /usr/local/projdata/0568/projects/PLANKTON/illumina_aallen/jespinoz/References/Cylindrotheca/annotations/cylindrotheca_rd3.all.maker.gff
##gff-version 3
000161F|arrow   .   contig  1   17082   .   .   .   ID=000161F|arrow;Name=000161F|arrow
000161F|arrow   maker   gene    35  1654    .   -   .   ID=CylG030441|arrow-snap-gene-0.5;Name=CylG030441|arrow-snap-gene-0.5;Alias=maker-000161F|arrow-snap-gene-0.5;
000161F|arrow   maker   mRNA    35  1654    .   -   .   ID=CylG030441|arrow-snap-gene-0.5-mRNA-1;Parent=CylG030441|arrow-snap-gene-0.5;Name=CylG030441|arrow-snap-gene-0.5-mRNA-1;Alias=maker-000161F|arrow-snap-gene-0.5-mRNA-1;_AED=0.13;_QI=0|0|0|1|0|0|2|0|510;_eAED=0.07;
000161F|arrow   maker   exon    1406    1654    .   -   .   ID=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1:exon:28;Parent=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1;
000161F|arrow   maker   exon    35  1315    .   -   .   ID=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1:exon:27;Parent=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1;
000161F|arrow   maker   CDS 1406    1654    .   -   0   ID=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1:cds;Parent=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1;
000161F|arrow   maker   CDS 35  1315    .   -   0   ID=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1:cds;Parent=CylT030441-RA|arrow-snap-gene-0.5-mRNA-1;
000161F|arrow   maker   gene    15768   16170   .   +   .   ID=CylG030449|arrow-snap-gene-0.4;Name=CylG030449|arrow-snap-gene-0.4;Alias=maker-000161F|arrow-snap-gene-0.4;
000161F|arrow   maker   mRNA    15768   16170   .   +   .   ID=CylG030449|arrow-snap-gene-0.4-mRNA-1;Parent=CylG030449|arrow-snap-gene-0.4;Name=CylG030449|arrow-snap-gene-0.4-mRNA-1;Alias=maker-000161F|arrow-snap-gene-0.4-mRNA-1;_AED=0.41;_QI=0|0|0|1|0|0|3|0|94;_eAED=0.41;
```

Lastly, here is a few lines of my converted gtf file:
```
-bash-4.1$ head /usr/local/projdata/0568/projects/PLANKTON/illumina_aallen/jespinoz/References/Cylindrotheca/annotations/cylindrotheca_rd3.all.maker.gtf
000000F|arrow   maker   exon    266 1689    .   +   .   transcript_id "CylT000001-RA|arrow-processed-gene-0.141-mRNA-1";
000000F|arrow   maker   exon    1781    2574    .   +   .   transcript_id "CylT000001-RA|arrow-processed-gene-0.141-mRNA-1";
000000F|arrow   maker   CDS 266 634 .   +   0   transcript_id "CylT000001-RA|arrow-processed-gene-0.141-mRNA-1";
000000F|arrow   maker   exon    326 473 .   +   .   transcript_id "CylT000002-RA|arrow-noncoding-gene-0.302-snoRNA-1";
000000F|arrow   maker   exon    2798    3631    .   -   .   transcript_id "CylT000003-RA|arrow-processed-gene-0.179-mRNA-1";
000000F|arrow   maker   CDS 2798    3631    .   -   0   transcript_id "CylT000003-RA|arrow-processed-gene-0.179-mRNA-1";
000000F|arrow   maker   exon    6084    6181    .   +   .   transcript_id "CylT000004-RA|arrow-noncoding-gene-0.303-snoRNA-1";
000000F|arrow   maker   exon    7306    9415    .   +   .   transcript_id "CylT000005-RA|arrow-snap-gene-0.2-mRNA-1";
000000F|arrow   maker   exon    9971    10218   .   +   .   transcript_id "CylT000005-RA|arrow-snap-gene-0.2-mRNA-1";
000000F|arrow   maker   CDS 7306    9415    .   +   0   transcript_id "CylT000005-RA|arrow-snap-gene-0.2-mRNA-1";
```

#### My questions:

* How can I structure my `gtf` file so it has everything necessary to run with this script? 

* In the `gtf` file does it only look for `CDS` and `exon` or does it want the `mRNA` and `gene` records as well?   

* Can I keep "|" characters or does this need to be modified.  

* Does every record, including exon, need a `transcript_id`, `gene_id`, `gene_name`, and `transcript_name`?

**Comment by alecw** [comment]
Note the [GTF format specification](https://useast.ensembl.org/info/website/upload/gff.html).

* How can I structure my gtf file so it has everything necessary to run with this script?
  * The attribute section needs to have all the fields mentioned in the error message.

* In the gtf file does it only look for CDS and exon or does it want the mRNA and gene records as well?
  * It needs gene, exon, CDS.  mRNA is not needed.

* Can I keep "|" characters or does this need to be modified.
  * The first tab-separated column of the GTF is the sequence name.  These need to match the sequences in your reference fasta.  If '000000F|arrow' is the name of a reference sequence, then this will work fine.  If not, you've got a problem.

* Does every record, including exon, need a transcript_id, gene_id, gene_name, and transcript_name?
  * You will get error messages for all the things that are needed but not present.

**Comment by naomi-croft** [comment]
Hi, I am also having problems with my gtf annotation file. Can someone suggest what to try next. I am using the pipefish annotation file, given as: syty_maker.putative_function.domain_added_maker.gff

I used miniconda to convert this file to a gtf file, using this command:
gffread syty_maker.putative_function.domain_added_maker.gff -T -o syty_maker.putative_function.domain_added_maker.gtf

When I run the TagReadWithGeneExon command as this:

TagReadWithGeneExon \
I=star_1-PB1_Aligned.out.sorted.retagged.sam \
O=star_1-PB1_Aligned.out.sorted.retagged_gene_exon_tagged.bam \
ANNOTATIONS_FILE=syty_maker.putative_function.domain_added_maker.gtf \
TAG=GE 2>star_1-PB1_exontagging.err

The next file is not made properly. This is the read out:

<img width="453" alt="image" src="https://user-images.githubusercontent.com/36161461/59023621-739de680-8850-11e9-8939-ad52838e4126.png">

**Comment by alecw** [comment]
Hi @naomi-croft ,

You are using an old version of the tools.  Get the latest release [here](https://github.com/broadinstitute/Drop-seq/releases/tag/v2.3.0) and try again.  I believe this is a bug that was fixed in release 2.1.0.

-Alec

P.S. If you still have problems, please create a new issue unless your problem is really related to this issue.
